src/Plugin/Field/FieldFormatter/SmartDocsFormatter.php

```diff
@@ -0,0 +1,163 @@
+use Drupal\Core\Serialization\Yaml;
+use Drupal\file\Plugin\Field\FieldFormatter\FileFormatterBase;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Drupal\Component\Serialization\Json;
+
+/**
+ * Plugin implementation of the SmartDocs OpenAPI spec formatter.
+ *
+ * @FieldFormatter(
+ *   id = "apigee_api_catalog_smartdocs",
+ *   label = @Translation("SmartDocs"),
+ *   description = @Translation("Formats OpenAPI specs with SmartDocs"),
+ *   field_types = {
+ *     "file"
+ *   }
+ * )
+ */
+class SmartDocsFormatter extends FileFormatterBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The logger factory.
+   *
+   * @var \Drupal\Core\Logger\LoggerChannelFactoryInterface
+   */
+  protected $loggerFactory;
+
+  /**
+   * Constructs a EntityReferenceEntityFormatter instance.
+   *
+   * @param string $plugin_id
+   *   The plugin_id for the formatter.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Field\FieldDefinitionInterface $field_definition
+   *   The definition of the field to which the formatter is associated.
+   * @param array $settings
+   *   The formatter settings.
+   * @param string $label
+   *   The formatter label display setting.
+   * @param string $view_mode
+   *   The view mode.
+   * @param array $third_party_settings
+   *   Any third party settings settings.
+   * @param \Drupal\Core\Logger\LoggerChannelFactoryInterface $loggerFactory
+   *   The logger factory.
+   */
+  public function __construct($plugin_id, $plugin_definition, FieldDefinitionInterface $field_definition, array $settings, $label, $view_mode, array $third_party_settings, LoggerChannelFactoryInterface $loggerFactory) {
+    parent::__construct($plugin_id, $plugin_definition, $field_definition, $settings, $label, $view_mode, $third_party_settings);
+    $this->loggerFactory = $loggerFactory->get('apigee_api_catalog');;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $plugin_id,
+      $plugin_definition,
+      $configuration['field_definition'],
+      $configuration['settings'],
+      $configuration['label'],
+      $configuration['view_mode'],
+      $configuration['third_party_settings'],
+      $container->get('logger.factory')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function viewElements(FieldItemListInterface $items, $langcode) {
+
+    $entity = $items->getEntity();
+    $entity_type = $entity->getEntityTypeId();
+
+    // The list of OpenAPI specs to pass to SmartDocs Angular app.
+    $openapi_files = [];
+
+    /** @var \Drupal\file\Entity\File $file */
+    foreach ($this->getEntitiesToView($items, $langcode) as $delta => $file) {
+      $file_extension = pathinfo($file->getFilename(), PATHINFO_EXTENSION);
+      $file_data = file_get_contents($file->getFileUri());
```

Comment:
This could be allot of data to load/parse to add server side validation using allot of memory. How does the client side app handle an invalid spec. It might be better to validate when the file is saved as opposed to every time it is rendered.

Reply:
The Angular app needs a javascript object, so we need to parse the YAML file into JSON for the Angular app to consume.  We could parse the YAML file in the Angular app after adding an appropriate YAML parser to the Angular, but after talking to SmartDocs devs they thought converting it beforehand would be better.  I tried not parsing JSON files and only YAML, but if you pass the raw file to drupalSettings, the file is treated as a string which makes quotes get double encoded when adding to session storage.  Validating valid YAML/JSON would be good during save also however, but even better would be to validate that it is a valid OpenAPI spec.

Reply:
Updated PR to have the field formatter put the file URL and file extension into drupalSettings instead of actual file.  The smartdos_integration.js script is now responsible for pulling file from server and parsing into javascript to save into session storage for SmartDocs Angular app to retrieve.

Reply:
Awesome!